r/src/compute.cpp

```diff
@@ -574,3 +576,89 @@ SEXP compute__CallFunction(std::string func_name, cpp11::list args, cpp11::list
 std::vector<std::string> compute__GetFunctionNames() {
   return arrow::compute::GetFunctionRegistry()->GetFunctionNames();
 }
+
+class RScalarUDFCallable : public arrow::compute::ArrayKernelExec {
+ public:
+  RScalarUDFCallable(const std::shared_ptr<arrow::Schema>& input_types,
+                     const std::shared_ptr<arrow::DataType>& output_type, cpp11::sexp fun)
+      : input_types_(input_types), output_type_(output_type), fun_(fun) {}
+
+  arrow::Status operator()(arrow::compute::KernelContext* context,
+                           const arrow::compute::ExecSpan& span,
+                           arrow::compute::ExecResult* result) {
+    auto fun_result = SafeCallIntoR<std::shared_ptr<arrow::Array>>([&]() {
+      cpp11::writable::list args_sexp;
+      args_sexp.reserve(span.num_values());
+
+      for (int64_t i = 0; i < span.num_values(); i++) {
+        const arrow::compute::ExecValue& v = span[i];
+        if (v.is_array()) {
+          std::shared_ptr<arrow::Array> array = v.array.ToArray();
+          args_sexp.push_back(cpp11::to_r6<arrow::Array>(array));
+        } else if (v.is_scalar()) {
+          std::shared_ptr<const arrow::Scalar> scalar = v.scalar->shared_from_this();
+          args_sexp.push_back(cpp11::to_r6<const arrow::Scalar>(scalar));
```

Comment:
```suggestion
          std::shared_ptr<arrow::Scalar> scalar = v.scalar->Copy();
          args_sexp.push_back(cpp11::to_r6<arrow::Scalar>(scalar));
```

Reply:
I compiled locally and ran the test cases, it works fine. But let me also try your example where this segfaults.

Reply:
And the output from the given test case is as follows after the modification

```R
> library(arrow, warn.conflicts = FALSE)
> fun <- arrow:::arrow_scalar_function(
+   int32(),
+   int64(),
+   function(x, y) {
+     y[[1]]$cast(int64())
+   }
+ )
> arrow:::register_scalar_function("my_test_scalar_function", fun)
> call_function("my_test_scalar_function", Array$create(1L))
Array
<int64>
[
  1
]
> call_function("my_test_scalar_function", Scalar$create(1L))
Error: arrow_scalar_function must return an Array
```

I think now it is a matter of outputting an `Array` instead of an `Scalar`

Reply:
Is making copies and passing values to the R function is a must?. @paleolimbot do you have an idea about that? I am not very much sure how R-CPP bindings work.